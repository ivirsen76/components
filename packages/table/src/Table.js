import React from 'react'
import PropTypes from 'prop-types'
import Paginator from '@ieremeev/paginator'
import md5 from 'md5'
import storage from 'store'
import _find from 'lodash/find'
import _filter from 'lodash/filter'
import _escapeRegExp from 'lodash/escapeRegExp'
import _sortBy from 'lodash/sortBy'
import _isEqual from 'lodash/isEqual'
import _pick from 'lodash/pick'
import style from './Table.module.scss'
import Filter from './Filter'
import Sortable from './Sortable'

/** Table component */
export default class Table extends React.Component {
    static propTypes = {
        /** unique name for the table */
        name: PropTypes.string,

        /** Class for the <table /> */
        className: PropTypes.string,

        /**
         * Array of columns. For example:
         * [
         *     { name: 'id', label: 'ID' },
         *     { name: 'name', label: 'Name', sort: true, filter: true },
         *     { name: 'email', label: 'Email', sort: true, filter: true },
         * ]
         */
        columns: PropTypes.array,

        /**
         * Array of data. For example:
         * [
         *     { id: 1, name: 'Mike', email: 'mike@gmail.com' },
         *     { id: 2, name: 'Helen', email: 'helen@gmail.com' },
         *     { id: 3, name: 'Bob', email: 'bob@gmail.com' },
         * ]
         */
        data: PropTypes.array,

        /** Show row number or not? */
        showRowNumber: PropTypes.bool,

        /** Default number of rows per one page */
        perPage: PropTypes.number,

        /** Show a paginator below the table or not? */
        hideBottomPaginator: PropTypes.bool,

        /** Message shown when there is no data to show */
        noDataMessage: PropTypes.string,
    }

    static defaultProps = {
        name: 'table',
        className: 'ui compact table',
        columns: [],
        data: [],
        showRowNumber: false,
        perPage: 50,
        hideBottomPaginator: false,
        noDataMessage: 'There is no data',
    }

    constructor(props) {
        super(props)

        this.state = {
            currentPage: 1,
            filters: {},
            orderBy: '',
            isAscentOrder: true,
            ...this.getSettings(),
        }
    }

    componentDidUpdate() {
        this.saveSettings()
    }

    getHash = () => {
        // Generate unique hash for the page+table
        let str =
            window.location.hostname +
            window.location.pathname +
            this.props.name +
            this.props.columns.map(column => column.name).toString()
        let hash = md5(str)

        return 'table_' + hash
    }

    getSettings = () => {
        let settings = storage.get(this.getHash())
        if (!settings) {
            return {}
        }

        return _pick(settings, ['filters', 'orderBy', 'isAscentOrder'])
    }

    saveSettings = () => {
        const hash = this.getHash()
        const oldSettings = storage.get(hash)
        const newSettings = _pick(this.state, ['filters', 'orderBy', 'isAscentOrder'])

        if (!_isEqual(oldSettings, newSettings)) {
            storage.set(hash, newSettings)
        }
    }

    getStartRowNumber = () => (this.state.currentPage - 1) * this.props.perPage

    getFilteredData = () => {
        let rows = this.props.data

        // Apply filters
        this.props.columns.forEach(column => {
            const filterValue = this.state.filters[column.name]
            if (!column.filter || !filterValue) {
                return
            }

            const regex = new RegExp(_escapeRegExp(filterValue), 'i')
            rows = _filter(rows, row => row[column.name] && row[column.name].match(regex))
        })

        // Apply an order
        if (this.state.orderBy !== '') {
            const column = _find(this.props.columns, { name: this.state.orderBy })

            // Check if the column hasn't been removed
            if (column) {
                rows = _sortBy(rows, this.state.orderBy)

                if (!this.state.isAscentOrder) {
                    rows.reverse()
                }
            }
        }

        return rows
    }

    getPageRows = () => {
        const start = this.getStartRowNumber()
        const end = this.state.currentPage * this.props.perPage

        return this.getFilteredData().slice(start, end)
    }

    getFilteredTotal = () => this.getFilteredData().length

    getTotal = () => this.props.data.length

    hasFilters = () => !!_find(this.props.columns, { filter: true })

    gotoPage = page => {
        this.setState({
            currentPage: page,
        })
    }

    onFilterChange = (column, value) => {
        const filters = this.state.filters
        if (value === '') {
            delete filters[column]
        } else {
            filters[column] = value
        }

        this.setState({ currentPage: 1, filters })
    }

    onOrder = orderBy => {
        const isAscentOrder =
            this.state.orderBy === orderBy ? !this.state.isAscentOrder : this.state.isAscentOrder

        this.setState({
            currentPage: 1,
            orderBy,
            isAscentOrder,
        })
    }

    renderPaginator = () => {
        const total = Math.ceil(this.getFilteredTotal() / this.props.perPage)

        if (total <= 1) {
            return null
        }

        return (
            <Paginator
                total={total}
                currentPage={this.state.currentPage}
                onPageChange={this.gotoPage}
            />
        )
    }

    renderThead = () => {
        const { columns, showRowNumber } = this.props

        const list = columns.map(column => {
            if (!column.sort) {
                return <th key={column.name}>{column.label}</th>
            }

            let sortable = <Sortable />
            if (column.name === this.state.orderBy) {
                sortable = this.state.isAscentOrder ? (
                    <Sortable order="asc" />
                ) : (
                    <Sortable order="desc" />
                )
            }

            return (
                <th
                    className={style.sortable}
                    key={column.name}
                    onClick={this.onOrder.bind(this, column.name)}
                >
                    {column.label}
                    {sortable}
                </th>
            )
        })

        // Is Number?
        if (showRowNumber) {
            list.unshift(
                <th key="number_column" className={style.numberColumn}>
                    #
                </th>
            )
        }

        return (
            <thead className={this.hasFilters() && style.hasFilters}>
                <tr>{list}</tr>
                {this.renderFilter()}
            </thead>
        )
    }

    renderFilter = () => {
        const { columns, showRowNumber } = this.props

        if (!this.hasFilters()) {
            return null
        }

        const filters = columns.map(column => {
            if (!column.filter) {
                return <th key={column.name} />
            }

            return (
                <th key={column.name}>
                    <Filter
                        column={column.name}
                        value={this.state.filters[column.name] || ''}
                        onFilterChange={this.onFilterChange}
                    />
                </th>
            )
        })

        if (showRowNumber) {
            filters.unshift(<th key="number_column" />)
        }

        return <tr>{filters}</tr>
    }

    renderTbody = () => {
        const { columns, showRowNumber, noDataMessage } = this.props
        const start = this.getStartRowNumber()

        let rows = this.getPageRows()
        if (rows.length === 0) {
            return (
                <tbody>
                    <tr>
                        <td colSpan={100}>{noDataMessage}</td>
                    </tr>
                </tbody>
            )
        }

        rows = rows.map((row, index) => {
            const list = columns.map(column => (
                <td key={column.name} className={column.className}>
                    {column.render ? column.render(row[column.name], row) : row[column.name]}
                </td>
            ))

            // Is Number?
            if (showRowNumber) {
                list.unshift(
                    <td key="number_column" className={style.numberColumn}>
                        {start + index + 1}
                    </td>
                )
            }

            return <tr key={row.id}>{list}</tr>
        })

        return <tbody>{rows}</tbody>
    }

    render() {
        const { className, hideBottomPaginator } = this.props
        const paginator = this.renderPaginator()

        return (
            <div>
                {paginator}
                <table className={className}>
                    {this.renderThead()}
                    {this.renderTbody()}
                </table>
                {!hideBottomPaginator && paginator}
            </div>
        )
    }
}
