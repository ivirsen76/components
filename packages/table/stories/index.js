import React from 'react'
import Table from '../src/Table'
import { storiesOf } from '@kadira/storybook' // eslint-disable-line
import style from './style.module.scss'

const data = [
    { id: 11, name: 'Mike', email: 'mike@gmail.com' },
    { id: 22, name: 'Helen', email: 'helen@gmail.com' },
    { id: 33, name: 'Bob', email: 'bob@gmail.com' },
]

const columns = [
    { name: 'id', label: 'ID' },
    { name: 'name', label: 'Name' },
    { name: 'email', label: 'Email' },
]

const stories = storiesOf('table', module)

stories.add('simple', () => <Table columns={columns} data={data} />)

stories.add('with no data', () => <Table columns={columns} data={[]} noDataMessage="No data" />)

stories.add('custom table class', () =>
    <Table className="ui striped celled table" columns={columns} data={data} />
)

stories.add('custom column render', () => {
    const adjustedColumns = [
        { name: 'id', label: 'ID' },
        {
            name: 'name',
            label: 'Name',
            render(value, row) {
                return (
                    <div className="ui label">
                        {value}
                    </div>
                )
            },
        },
        {
            name: 'email',
            label: 'Email',
            render(value, row) {
                return (
                    <a href={`email:${value}`}>
                        {value}
                    </a>
                )
            },
        },
    ]

    return <Table columns={adjustedColumns} data={data} />
})

stories.add('show row number', () => <Table columns={columns} data={data} showRowNumber />)

stories.add('with pagination', () =>
    <Table columns={columns} data={data} perPage={2} showRowNumber />
)

stories.add('hide bottom paginator', () =>
    <Table columns={columns} data={data} perPage={2} hideBottomPaginator />
)

stories.add('show filters', () => {
    const adjustedColumns = [
        { name: 'id', label: 'ID' },
        { name: 'name', label: 'Name', filter: true },
        { name: 'email', label: 'Email', filter: true },
    ]

    return <Table columns={adjustedColumns} data={data} />
})

stories.add('show sorting', () => {
    const adjustedColumns = [
        { name: 'id', label: 'ID' },
        { name: 'name', label: 'Name', sort: true },
        { name: 'email', label: 'Email', sort: true },
    ]

    return <Table columns={adjustedColumns} data={data} />
})

stories.add('with sort, filter, row number', () => {
    const adjustedColumns = [
        { name: 'id', label: 'ID' },
        { name: 'name', label: 'Name', filter: true, sort: true },
        { name: 'email', label: 'Email', filter: true, sort: true },
    ]

    return <Table columns={adjustedColumns} data={data} showRowNumber />
})

stories.add('with custom class for column', () => {
    const adjustedColumns = [
        { name: 'id', label: 'ID' },
        { name: 'name', label: 'Name' },
        { name: 'email', label: 'Email', className: style.nowrap },
    ]
    const newData = [
        {
            id: 44,
            name: 'Word '.repeat(40),
            email: "this one shouldn't be wrapped",
        },
    ]

    return <Table columns={adjustedColumns} data={newData} />
})
