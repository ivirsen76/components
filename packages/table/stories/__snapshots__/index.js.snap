// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`table custom column render 1`] = `
<div>
  <table
    className="ui compact table"
  >
    <thead
      className={false}
    >
      <tr>
        <th>
          ID
        </th>
        <th>
          Name
        </th>
        <th>
          Email
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          11
        </td>
        <td>
          <div
            className="ui label"
          >
            Mike
          </div>
        </td>
        <td>
          <a
            href="email:mike@gmail.com"
          >
            mike@gmail.com
          </a>
        </td>
      </tr>
      <tr>
        <td>
          22
        </td>
        <td>
          <div
            className="ui label"
          >
            Helen
          </div>
        </td>
        <td>
          <a
            href="email:helen@gmail.com"
          >
            helen@gmail.com
          </a>
        </td>
      </tr>
      <tr>
        <td>
          33
        </td>
        <td>
          <div
            className="ui label"
          >
            Bob
          </div>
        </td>
        <td>
          <a
            href="email:bob@gmail.com"
          >
            bob@gmail.com
          </a>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`table custom table class 1`] = `
<div>
  <table
    className="ui striped celled table"
  >
    <thead
      className={false}
    >
      <tr>
        <th>
          ID
        </th>
        <th>
          Name
        </th>
        <th>
          Email
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          11
        </td>
        <td>
          Mike
        </td>
        <td>
          mike@gmail.com
        </td>
      </tr>
      <tr>
        <td>
          22
        </td>
        <td>
          Helen
        </td>
        <td>
          helen@gmail.com
        </td>
      </tr>
      <tr>
        <td>
          33
        </td>
        <td>
          Bob
        </td>
        <td>
          bob@gmail.com
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`table hide bottom paginator 1`] = `
<div>
  <Paginator
    aroundPages={3}
    currentPage={1}
    edgePages={3}
    isEdgeLinks={true}
    minPages={10}
    onPageChange={[Function]}
    showAngularLinks={true}
    total={2}
  />
  <table
    className="ui compact table"
  >
    <thead
      className={false}
    >
      <tr>
        <th>
          ID
        </th>
        <th>
          Name
        </th>
        <th>
          Email
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          11
        </td>
        <td>
          Mike
        </td>
        <td>
          mike@gmail.com
        </td>
      </tr>
      <tr>
        <td>
          22
        </td>
        <td>
          Helen
        </td>
        <td>
          helen@gmail.com
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`table show filters 1`] = `
<div>
  <table
    className="ui compact table"
  >
    <thead
      className="hasFilters"
    >
      <tr>
        <th>
          ID
        </th>
        <th>
          Name
        </th>
        <th>
          Email
        </th>
      </tr>
      <tr>
        <th />
        <th>
          <Filter
            column="name"
            onFilterChange={[Function]}
            value=""
          />
        </th>
        <th>
          <Filter
            column="email"
            onFilterChange={[Function]}
            value=""
          />
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          11
        </td>
        <td>
          Mike
        </td>
        <td>
          mike@gmail.com
        </td>
      </tr>
      <tr>
        <td>
          22
        </td>
        <td>
          Helen
        </td>
        <td>
          helen@gmail.com
        </td>
      </tr>
      <tr>
        <td>
          33
        </td>
        <td>
          Bob
        </td>
        <td>
          bob@gmail.com
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`table show row number 1`] = `
<div>
  <table
    className="ui compact table"
  >
    <thead
      className={false}
    >
      <tr>
        <th
          className="numberColumn"
        >
          #
        </th>
        <th>
          ID
        </th>
        <th>
          Name
        </th>
        <th>
          Email
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="numberColumn"
        >
          1
        </td>
        <td>
          11
        </td>
        <td>
          Mike
        </td>
        <td>
          mike@gmail.com
        </td>
      </tr>
      <tr>
        <td
          className="numberColumn"
        >
          2
        </td>
        <td>
          22
        </td>
        <td>
          Helen
        </td>
        <td>
          helen@gmail.com
        </td>
      </tr>
      <tr>
        <td
          className="numberColumn"
        >
          3
        </td>
        <td>
          33
        </td>
        <td>
          Bob
        </td>
        <td>
          bob@gmail.com
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`table show sorting 1`] = `
<div>
  <table
    className="ui compact table"
  >
    <thead
      className={false}
    >
      <tr>
        <th>
          ID
        </th>
        <th
          className="sortable"
          onClick={[Function]}
        >
          Name
          <Sortable />
        </th>
        <th
          className="sortable"
          onClick={[Function]}
        >
          Email
          <Sortable />
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          11
        </td>
        <td>
          Mike
        </td>
        <td>
          mike@gmail.com
        </td>
      </tr>
      <tr>
        <td>
          22
        </td>
        <td>
          Helen
        </td>
        <td>
          helen@gmail.com
        </td>
      </tr>
      <tr>
        <td>
          33
        </td>
        <td>
          Bob
        </td>
        <td>
          bob@gmail.com
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`table simple 1`] = `
<div>
  <table
    className="ui compact table"
  >
    <thead
      className={false}
    >
      <tr>
        <th>
          ID
        </th>
        <th>
          Name
        </th>
        <th>
          Email
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          11
        </td>
        <td>
          Mike
        </td>
        <td>
          mike@gmail.com
        </td>
      </tr>
      <tr>
        <td>
          22
        </td>
        <td>
          Helen
        </td>
        <td>
          helen@gmail.com
        </td>
      </tr>
      <tr>
        <td>
          33
        </td>
        <td>
          Bob
        </td>
        <td>
          bob@gmail.com
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`table with custom class for column 1`] = `
<div>
  <table
    className="ui compact table"
  >
    <thead
      className={false}
    >
      <tr>
        <th>
          ID
        </th>
        <th>
          Name
        </th>
        <th>
          Email
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          44
        </td>
        <td>
          Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word Word 
        </td>
        <td
          className="nowrap"
        >
          this one shouldn't be wrapped
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`table with no data 1`] = `
<div>
  <table
    className="ui compact table"
  >
    <thead
      className={false}
    >
      <tr>
        <th>
          ID
        </th>
        <th>
          Name
        </th>
        <th>
          Email
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          colSpan={100}
        >
          No data
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`table with pagination 1`] = `
<div>
  <Paginator
    aroundPages={3}
    currentPage={1}
    edgePages={3}
    isEdgeLinks={true}
    minPages={10}
    onPageChange={[Function]}
    showAngularLinks={true}
    total={2}
  />
  <table
    className="ui compact table"
  >
    <thead
      className={false}
    >
      <tr>
        <th
          className="numberColumn"
        >
          #
        </th>
        <th>
          ID
        </th>
        <th>
          Name
        </th>
        <th>
          Email
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="numberColumn"
        >
          1
        </td>
        <td>
          11
        </td>
        <td>
          Mike
        </td>
        <td>
          mike@gmail.com
        </td>
      </tr>
      <tr>
        <td
          className="numberColumn"
        >
          2
        </td>
        <td>
          22
        </td>
        <td>
          Helen
        </td>
        <td>
          helen@gmail.com
        </td>
      </tr>
    </tbody>
  </table>
  <Paginator
    aroundPages={3}
    currentPage={1}
    edgePages={3}
    isEdgeLinks={true}
    minPages={10}
    onPageChange={[Function]}
    showAngularLinks={true}
    total={2}
  />
</div>
`;

exports[`table with sort, filter, row number 1`] = `
<div>
  <table
    className="ui compact table"
  >
    <thead
      className="hasFilters"
    >
      <tr>
        <th
          className="numberColumn"
        >
          #
        </th>
        <th>
          ID
        </th>
        <th
          className="sortable"
          onClick={[Function]}
        >
          Name
          <Sortable />
        </th>
        <th
          className="sortable"
          onClick={[Function]}
        >
          Email
          <Sortable />
        </th>
      </tr>
      <tr>
        <th />
        <th />
        <th>
          <Filter
            column="name"
            onFilterChange={[Function]}
            value=""
          />
        </th>
        <th>
          <Filter
            column="email"
            onFilterChange={[Function]}
            value=""
          />
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          className="numberColumn"
        >
          1
        </td>
        <td>
          11
        </td>
        <td>
          Mike
        </td>
        <td>
          mike@gmail.com
        </td>
      </tr>
      <tr>
        <td
          className="numberColumn"
        >
          2
        </td>
        <td>
          22
        </td>
        <td>
          Helen
        </td>
        <td>
          helen@gmail.com
        </td>
      </tr>
      <tr>
        <td
          className="numberColumn"
        >
          3
        </td>
        <td>
          33
        </td>
        <td>
          Bob
        </td>
        <td>
          bob@gmail.com
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
